README
# README (this file) FlowViewer V4.6  Date: 01/26/2015
#
# FlowViewer is a set of three tools (FlowViewer, FlowGrapher,
# FlowMonitor) that create text reports, graph reports, and 
# long-term monitor reports from flow-tools and SiLK captured
# and stored netflow data. FlowViewer can run with both flow-tools
# and SiLK simultaneously. Flow-tools can handle up to v7; SilK 
# can handle v5, v9, and IPFIX. The User's Guide is very helpful.
#
# Software Dependencies:
# 
#   flow-tools  http://code.google.com/p/flow-tools (If collecting v5 only)
#   SiLK        http://tools.netsa.cert.org/silk    (If collecting IPFIX)
#   libfixbuf   http://tools.netsa.cert.org/silk    (If collecting IPFIX)
#   gd          http://www.libgd.org/Downloads
#   GD          http://search.cpan.org/~lds/GD-2.30
#   GD::Graph   http://search.cpan.org/~mverb/GDGraph-1.43
#   GD::Text    http://search.cpan.org/~mverb/GDTextUtil-0.86/Text
#   RRDtool     http://people.ee.ethz.ch/~oetiker/webtools/rrdtool/pub
#
# Quick Upgrade
#
#   0. If using SiLK, must upgrade to v3.8.0 or newer
#   1. Untar the package into a new cgi-bin subdirectory
#   2. Configure FlowViewer_Configuration.pm variables to your environment
#      and create all necessary directories with proper permissions
#   3. Replace old logos with new buttons (will be done automatically)
#   4. Copy FlowViewer.css, FlowViewer.pdf to $reports_directory
#   5. Configure FlowViewer_Configuration.pm to point to existing 
#      FlowMonitor_Filter and FlowMonitor_RRDtool directories
#      [For prior v4.5 you can rename these directories to:
#      FlowMonitor_Filter and FlowMonitor_RRDtool or change the 
#      configuration variables to point to the old directories.]
#   6. Configure new FlowViewer_Configuration.pm
#   7. Stop old FlowMonitor_Collector and FlowMonitor_Grapher
#      [Upgrading from v4.4 and prior you will be stopping FlowTracker_Collector
#      and FlowTracker_Grapher.]
#   8. Start new FlowMonitor_Collector and FlowMonitor_Grapher
#   9. Copy NamedInterfaces_Devices, names file, user logo to new directory
#  10. (If upgrading from pre v4.0) Run convert_pre40_filters against existing filters 
#      (ie FlowViewer_SavedFilters)
#  11. Use included 'User Relay' scripts if desired (recommended - see below)
#
# Quick Install
#
#   1. Untar into cgi-bin subdirectory
#
#   For netflow v5 and older (option):
#
#   2. Download, install, configure flow-tools
#
#   For IPFIX (e.g., v9 - also handles v5):
#
#   3. Download, install, configure SiLK (v3.8.0 or newer) and libfixbuf
#
#   For sflow
#
#   4. From SiLK FAQ:
#
#      "Support for sFlow v5 is available as of SiLK 3.9.0 when you configure
#      and build SiLK to use v1.6.0 or later of the libfixbuf library."
#
#   For FlowViewer
#
#   5. Configure FlowViewer_Configuration.pm variables as necessary
#   6. Create all necessary directories with proper permissions
#   7. Copy FlowViewer.css, FlowViewer.pdf to $reports_directory
#   8. Point browser to FV.cgi
#
#   For FlowGrapher
#
#   9. Install gd (C), GD (Perl), GD::Graph (Perl) GD::Text (Perl)
#  10. Configure FlowViewer_Configuration.pm variables as necessary
#  11. Point browser to FV.cgi
#
#   For FlowMonitor
#
#  12. Install RRDtool (at least version 1.4)
#  13. Create FlowMonitor_Filter and FlowMonitor_RRDtool directories
#  14. Configure FlowViewer_Configuration.pm variables as necessary
#  15. Start FlowMonitor_Collector, FlowMonitor_Grapher in background
#  16. Point browser to FV.cgi
#
#   For all FlowViewer tools
#  
#  17. Review all FlowViewer directories and files for proper permissions
#
# Version 4.6 Release Notes
#
#  Version 4.6 fixes local timezone difficulties that were not fixed as
#  advertised in version 4.5 for FlowGrapher and FlowViewer. Thanks goes
#  to Randy Feeney. Also note that version 4.6 removes the "$time_zone"
#  configurable parameter from FlowViewer_Configuration. Timezone is now
#  exclusively extracted from the system, using the 'date' function. This
#  version fixes a problem with FlowGrapher not correctly displaying the
#  smallest flows when requested (e.g., Detail Lines: -100 for smallest
#  100 flows.) Fixes improper listing of very old Saved files.
#
# Version 4.5 Release Notes
#
#  Version 4.5 resolves an unfortunate name clash in commercial space and
#  renames FlowTracker to FlowMonitor. The situation where SiLK data is
#  saved in UTC (GMT) time, but the system is left in local time has been
#  fixed (thanks to Kees Leune.) A new configuration variable
#  "$silk_compiled_localtime" has been added for the environment where SiLK
#  has been comipled with the --enable-localtime switch. FlowGrapher_Analyze
#  has been fixed to handle hyper-links to IPv6 hosts properly. SiLK IPsets
#  can now be input through the various tool menus. A problem with
#  multi-word Dashboards and Group creation has been fixed. Corrected
#  flows/second initiated calculation. Added the ability to bypass the
#  printing of pulldowns on the bottom service bar. Fixed an error with 
#  filtering on port equal to '0'. Fixed 'Len' field output for some
#  FlowGrapher reports. New parameter: $ipfix_default_device allows IPFIX
#  users to pre-select a primary device (e.g., using one sensor only.)
#  Extended pie-charts to some Printed reports. A new parameter
#  $site_config_file is added to make it easier to accomodate various
#  SiLK stored data file structures.
#
#  New FlowViewer_Configuration.pm parameters in v4.5:
#
#  $silk_compiled_localtime   - "Y" if SiLK compiled with local timezone
#  $ipset_directory           - Directory where IPsets can be found
#  $use_bottom_pulldowns      - Will exclude pulldowns on bottom of UI
#  $ipfix_default_device      - Controls the default in device_name pulldown
#  $sensor_config_file        - Changed from $sensor_config_directory
#  $site_config_file          - Left blank (= "";) will look in rootdir
#
#  Note: the rename of FlowTracker to FlowMonitor includes default names
#  for FlowMonitor related directories. The defaults that will previal if
#  no changes are made are:
#
#  $monitor_directory       = "/var/www/html/FlowMonitor";
#  $monitor_short           = "/FlowMonitor";
#  $filter_directory        = " ... /FlowMonitor_Files/FlowMonitor_Filters";  
#  $rrdtool_directory       = " ... /FlowMonitor_Files/FlowMonitor_RRDtool";
#
#  For users who are upgrading, these can be revised back to 'FlowTracker'
#  (or whatever) with no problem. The alternative is to simply rename the
#  existing directories.
#
# Version 4.4 Release Notes
#
#  Version 4.4 introduces two new significant capabilities; multiple
#  dashboards and FlowGrapher Analysis. The user can now set up 
#  more than one active dashboard with links on every page to get
#  to each dashboard. Uses include multiple networks, data centers,
#  or multiple users. The new version also permits users to quickly 
#  de-aggregate FlowGraphs into the largest (3 to 10) source or
#  destination IP or Port contributors. This is particularly useful in
#  analyzing peaks in graphs based on "Flows" for DDOS and the like.
#  The multiple dashboard changes touched most of the scripts. There
#  is one new script, FlowGrapher_Analyze, and the script
#  FlowMonitor_AltDashboard is deprecated. The code was cleansed of
#  confusing 'prorated (_P)' distinctions. Version 4.4 also greatly
#  improves the user's ability to interface with various SiLK
#  configurations other than the 'flow-tools-like' one suggested in
#  earlier documentation. Note: SiLK users must upgrade SiLK to 
#  version 3.8.0 or newer.
#
# Version 4.3 Release Notes
#
#  Version 4.3 introduces a new FlowViewer report called "Detect
#  Scanning" which employs scanning detection software in both
#  flow-tools (flow-dscan) and SiLK (rwscan.) The output from each 
#  report includes links to the individual scanner sources detected.
#  The link will create a FlowGrapher report for that host. Modified
#  field checks to permit 32 bit AS entries. However, this is for the
#  future when SiLK begins to handle the fields. Thanks Veasna Long.
#
# Version 4.2.2 Release Notes
#
#  Version 4.2.2 is a quick emergency fix with limited changes. An
#  array (@temp_ports) in FlowViewer_Utilities was not initialized
#  and was acumulating contents over many FlowMonitor_Collector runs.
#  This was causing the whole FlowMonitor_Collector run to slow down
#  gradually over time as this array would have to get sorted for each
#  SiLK FlowMonitor (this caused no problems for flow-tools only 
#  users.)
#
# Version 4.2.1 Release Notes
#
#  Version 4.2.1 extends the use of the recently discovered flow-report
#  option "linear-interpolated-flows-octets-packets" to FlowMonitor 
#  processing (see User's Guide for further discussion.) This provides 
#  significant speed-ups for FlowMonitor_Recreate (above 75%) and 
#  FlowMonitor_Collector (which will now permit many more FlowMonitors.)
#  The FlowMonitor_Recreate processing for SiLK was remarkably poorly
#  implemented and this has been corrected introducing unspeakable
#  performance gains there as well. The same use of SiLK prefiltering
#  that was introduced for FlowMonitor in v4.2 has been applied to
#  FlowGrapher for a speed-up there. A new capability is added to work
#  with different international date formats. The distribution includes a
#  new capability, FlowViewer_CleanSiLK, to monitor and adjust diskspace
#  used by IPFIX devices addressing the SiLK deficiency of not having the
#  valuable flow-capture feature (-E) of active diskspace usage control.
#  Made the use of prefiltered files or CONCATs non-optional and fixed
#  SiLK processing of flows and packets for FlowMonitor_Collector. Modest
#  changes have been made to FlowViewer_CleanFiles, FV.cgi and
#  flowmonitor_grapher_nonlazy. An error causing packets to be monitored as
#  flows in some cases was fixed. A new tool, flowmonitor_grapher_recent
#  will re-graph (nonlazy) only recently created FlowMonitors.
#
# Version 4.2 Release Notes
#
#  Version 4.2 incorporates the equivalent of "use existing concatenations"
#  for SiLK based FlowMonitors. FlowViewer takes advantage of previously
#  concatenated flow-tools files during FlowMonitor_Collector processing
#  to speed up the whole run. Now it does this for SiLK files as well by
#  performing rwfilter "INPUT" filtering only once for all FlowMonitors
#  that are based on the same sensor/class combination. This is a pretty
#  significant speed-up and will permit the user to have many more 
#  FlowMonitors. For example, our implementation, a combination of 
#  flow-tools and SiLK based data, now processes 250 FlowMonitors in 35
#  seconds (prior to the new version this was taking 50 seconds.) The new
#  version also corrects processing of exporters [M. Donnelly]. Excluded
#  fields (e.g., protocols=-17) were being accepted for SiLK FlowMonitors
#  despite SiLK not being able to handle them; this was fixed. The 
#  analyze_netflow_packets tool has been fixed for IPv6 addresses. Four new
#  date conversion utilities have been added to the 'tools' subdirectory.
#  FlowMonitor_Collector was modified in version 4.0 to be able to create
#  FlowMonitors for flows and packets (as well as bits.) This was
#  inadvertantly dropped in version 4.1 and is restored in this update 
#  to version 4.2 [7/31].
#
# Version 4.1 Release Notes
#
#  Version 4.1 includes a new FlowGrapher capability that creates reports
#  3 to 4 times faster than previously. The detail lines are a little bit 
#  different. The previous capability is retained offering the user a choice
#  from the input form interface. The new FlowGrapher report type is 
#  "Aggregated". It makes use of a heretofore missed flow-tools capability
#  known as the flow-report "linear-interpolated-flows-octets-packets" option
#  which aggregates flows, octets, or packets into time buckets. This moves
#  that processing into the compiled "C" code of flow-tools. New FlowMonitor
#  capabilities are added to monitor flows or packets as well as the previously
#  available octets. Version 4.1 introduces the ability to maintain different
#  dashboards for different users (please see the User's Guide for how to do
#  this. The new version includes a new FV_Relay.cgi script. The new version
#  fixes a flaw in FlowMonitor_Collector that erroneously monitored protocols,
#  tcp_flags, and tos_fields when using SiLK (thanks C. Spitzlay.) It also
#  includes some small fixes like making directory creation a little easier,
#  fixes removing (and adding) Trackings from the Dashboard and removes some
#  minor extraneous formatting. Fixed problem with FlowMonitor_Collector
#  processing of SiLK interface filtering [07/09/13]. Fixed initialization
#  of @ipfix_devices in FlowViewer_Configuration.pm [Thanks M. Donnelly.]
#
# Version 4.0 Release Notes
#
#  Version 4.0 is a major upgrade that enables FlowViewer to handle IPFIX 
#  netflow data (i.e., v9, etc.) The User Interface has been completely redone
#  and now features a Dashboard. Aside from the new collector interface and user
#  interface, version 4.0 introduces some new capabilities:
# 
#   1. FlowViewer report sorting by column header
#   2. Dashboard of thumbnail versions of selected FlowMonitor graphs
#   3. Ability to 'recreate' FlowMonitors, starting at a time specified in the 
#      past
#
#  The distribution manifest has changed significantly.
#
#  Preserved Scripts, Files, and Tools:
#
#   FlowViewer.cgi                Modified for new user interface.
#   FlowViewer_Main.cgi           Modified for new user interface.
#   FlowViewer_Relay.cgi          No change.
#   FlowViewer_Save.cgi           Significant modification.
#   FlowGrapher.cgi               Modified for new user interface.
#   FlowGrapher_Main.cgi          Modified for new interface.
#   FlowGrapher_Colors            No change.
#   FlowGrapher_Relay.cgi         No change.
#   FlowGrapher_Sort.cgi          Significant modification.
#   FlowMonitor.cgi               Modified for new user interface.
#   FlowMonitor_Collector         Modified to process stored SiLK data.
#   FlowMonitor_Grapher           Modified to update Thumbnails.
#   FlowMonitor_Group             Modified for new user interface.
#   FlowMonitor_Dumper            Modified for new user interface.
#   FlowMonitor_Relay.cgi         No change.
#   FlowViewer_CleanASCache       No change.
#   FlowViewer_CleanFiles         Minor changes.
#   FlowViewer_CleanHostCache     No change.
#   FlowViewer_Configuration.pm   Modifications for SiLK and user interface.
#   FlowViewer_Utilities.pm       Removed filter output processing.
#   NamedInterfaces_Devices       No change.
#   NamedInterfaces_Exporters     No change.
#   flowcapture_restart           No change.
#   flow-capture-table.conf       No change.
#   flowmonitor_restart           No change.
#   performance_check             Parse FlowMonitor logs and report performance
#   rsync_flows                   Rsync all of raw flow data to backup host
#   rsync_monitors               Rsync all of Tracking data to backup host
#
#  New Scripts, Files, and Tools
#
#   FlowViewer_Replay.cgi         Presents saved FlowViewer reports
#   FlowViewer_SaveManage.cgi     Manages saved reports
#   FlowViewer_Sort.cgi           Sorts FlowViewer reports
#   FlowViewer_UI.cgi             Utilities for creating user interface
#   FlowGrapher_Replay.cgi        Presents saved FlowGrapher reports
#   FlowMonitor_Dashboard.cgi     Manages the Dashboard contents
#   FlowMonitor_Display.cgi       Presents a FlowMonitor
#   FlowMonitor_DisplayPublic.cgi Presents a FlowMonitor from Public list
#   FlowMonitor_Management.cgi    Manages FlowMonitors (e.g., remove, etc.)
#   FlowMonitor_Recreate          Background process to recreate FlowMonitors
#   FlowMonitor_Thumbnail         Invoked to create a Thumbnail FlowMonitor
#   FlowViewer.css                FlowViewer cascading style sheet
#   FV_button.png                 New button link to FlowViewer from front page
#   FG_button.png                 New button link to FlowGrapher from front page
#   FM_button.png                 New button link to FlowMonitor from front page
#   convert_pre40_filters         Converts old saved filters (pre version 4.0).
#   flowmonitor_archive_restore   Restores archived FlowMonitors gone astray
#   flowmonitor_grapher_nonlazy   Forces a re-graphing of all FlowMonitor graphs
#   resize_rrdtools               Extends RRDtools created prior to 3-Year graph
#   rwflowpack_start              One-line script starts SiLK collector
#   analyze_netflow_packets       Script analyzes TCPDUMP captured netflow data 
#
#  Removed Scripts and Files
#

CHANGELOG.md

# Version 4.6 Release Notes

  Version 4.6 fixes local timezone difficulties that were not fixed as
  advertised in version 4.5 for FlowGrapher and FlowViewer. Thanks goes
  to Randy Feeney. Also note that version 4.6 removes the "$time_zone"
  configurable parameter from FlowViewer_Configuration. Timezone is now
  exclusively extracted from the system, using the 'date' function. This
  version fixes a problem with FlowGrapher not correctly displaying the
  smallest flows when requested (e.g., Detail Lines: -100 for smallest
  100 flows.) Fixes improper listing of very old Saved files.

# Version 4.5 Release Notes

  Version 4.5 resolves an unfortunate name clash in commercial space and
  renames FlowTracker to FlowMonitor. The situation where SiLK data is
  saved in UTC (GMT) time, but the system is left in local time has been
  fixed (thanks to Kees Leune.) A new configuration variable
  "$silk_compiled_localtime" has been added for the environment where SiLK
  has been comipled with the --enable-localtime switch. FlowGrapher_Analyze
  has been fixed to handle hyper-links to IPv6 hosts properly. SiLK IPsets
  can now be input through the various tool menus. A problem with
  multi-word Dashboards and Group creation has been fixed. Corrected
  flows/second initiated calculation. Added the ability to bypass the
  printing of pulldowns on the bottom service bar. Fixed an error with 
  filtering on port equal to '0'. Fixed 'Len' field output for some
  FlowGrapher reports. New parameter: $ipfix_default_device allows IPFIX
  users to pre-select a primary device (e.g., using one sensor only.)
  Extended pie-charts to some Printed reports. A new parameter
  $site_config_file is added to make it easier to accomodate various
  SiLK stored data file structures.

  New FlowViewer_Configuration.pm parameters in v4.5:

  $silk_compiled_localtime   - "Y" if SiLK compiled with local timezone
  $ipset_directory           - Directory where IPsets can be found
  $use_bottom_pulldowns      - Will exclude pulldowns on bottom of UI
  $ipfix_default_device      - Controls the default in device_name pulldown
  $sensor_config_file        - Changed from $sensor_config_directory
  $site_config_file          - Left blank (= "";) will look in rootdir

  Note: the rename of FlowTracker to FlowMonitor includes default names
  for FlowMonitor related directories. The defaults that will previal if
  no changes are made are:

  $monitor_directory       = "/var/www/html/FlowMonitor";
  $monitor_short           = "/FlowMonitor";
  $filter_directory        = " ... /FlowMonitor_Files/FlowMonitor_Filters";  
  $rrdtool_directory       = " ... /FlowMonitor_Files/FlowMonitor_RRDtool";

  For users who are upgrading, these can be revised back to 'FlowTracker'
  (or whatever) with no problem. The alternative is to simply rename the
  existing directories.

# Version 4.4 Release Notes

  Version 4.4 introduces two new significant capabilities; multiple
  dashboards and FlowGrapher Analysis. The user can now set up 
  more than one active dashboard with links on every page to get
  to each dashboard. Uses include multiple networks, data centers,
  or multiple users. The new version also permits users to quickly 
  de-aggregate FlowGraphs into the largest (3 to 10) source or
  destination IP or Port contributors. This is particularly useful in
  analyzing peaks in graphs based on "Flows" for DDOS and the like.
  The multiple dashboard changes touched most of the scripts. There
  is one new script, FlowGrapher_Analyze, and the script
  FlowMonitor_AltDashboard is deprecated. The code was cleansed of
  confusing 'prorated (_P)' distinctions. Version 4.4 also greatly
  improves the user's ability to interface with various SiLK
  configurations other than the 'flow-tools-like' one suggested in
  earlier documentation. Note: SiLK users must upgrade SiLK to 
  version 3.8.0 or newer.

# Version 4.3 Release Notes

  Version 4.3 introduces a new FlowViewer report called "Detect
  Scanning" which employs scanning detection software in both
  flow-tools (flow-dscan) and SiLK (rwscan.) The output from each 
  report includes links to the individual scanner sources detected.
  The link will create a FlowGrapher report for that host. Modified
  field checks to permit 32 bit AS entries. However, this is for the
  future when SiLK begins to handle the fields. Thanks Veasna Long.

# Version 4.2.2 Release Notes

  Version 4.2.2 is a quick emergency fix with limited changes. An
  array (@temp_ports) in FlowViewer_Utilities was not initialized
  and was acumulating contents over many FlowMonitor_Collector runs.
  This was causing the whole FlowMonitor_Collector run to slow down
  gradually over time as this array would have to get sorted for each
  SiLK FlowMonitor (this caused no problems for flow-tools only 
  users.)

# Version 4.2.1 Release Notes

  Version 4.2.1 extends the use of the recently discovered flow-report
  option "linear-interpolated-flows-octets-packets" to FlowMonitor 
  processing (see User's Guide for further discussion.) This provides 
  significant speed-ups for FlowMonitor_Recreate (above 75%) and 
  FlowMonitor_Collector (which will now permit many more FlowMonitors.)
  The FlowMonitor_Recreate processing for SiLK was remarkably poorly
  implemented and this has been corrected introducing unspeakable
  performance gains there as well. The same use of SiLK prefiltering
  that was introduced for FlowMonitor in v4.2 has been applied to
  FlowGrapher for a speed-up there. A new capability is added to work
  with different international date formats. The distribution includes a
  new capability, FlowViewer_CleanSiLK, to monitor and adjust diskspace
  used by IPFIX devices addressing the SiLK deficiency of not having the
  valuable flow-capture feature (-E) of active diskspace usage control.
  Made the use of prefiltered files or CONCATs non-optional and fixed

FlowViewer_Configuration.pm
#! /usr/bin/perl
#
#  Purpose:  
#  FlowViewer_Configuration.pm holds global variables for the FlowViewer,
#  FlowGrapher, and FlowMonitor NetFlow analysis tools.
#
#  Description:
#  Various parameters used to configure the system to a local environment.
#
#  Input arguments:
#  Name                 Description
#  -----------------------------------------------------------------------
#  None
#
#  Modification history:
#  Author       Date            Vers.   Description
#  -----------------------------------------------------------------------
#  J. Loiacono  07/04/2005      1.0     Original version.
#  J. Loiacono  01/01/2006      2.0     Flowgrapher, new functions, speed
#  J. Loiacono  01/16/2006      2.1     Introduced $flow_file_length
#  J. Loiacono  07/04/2006      3.0     Added parameters for Monitor, others
#  J. Loiacono  12/25/2006      3.1     Changes for MIN/MAX, permissions
#  J. Loiacono  02/22/2007      3.2     Minor changes for Groups
#  J. Loiacono  12/07/2007      3.3     Exporters, Logging, Time-zone
#                                       File cleanup, Pie charts
#  J. Loiacono  12/15/2007      3.3.1   New $no_devices ... parameter
#  J. Loiacono  09/11/2010      3.4     New default_report parameter
#  J. Loiacono  03/17/2011      3.4     Can now specify deafult report for FV
#                                       Host column widths are now adjutable
#                                       Added 'tries=1' to dig; speeds resolving
#  J. Loiacono  05/21/2011      3.4     Fixed speeded-up FlowGrapher for non-GMT hosts
#  J. Loiacono  05/08/2012      4.0     Major upgrade for IPFIX/v9 using SiLK,
#                                       New User Interface
#  J. Loiacono  04/15/2013      4.1     New default FlowGrapher report setting
#                                       Fixed @ipfix_devices variable [M. Donnelly]
#  J. Loiacono  09/11/2013      4.2.1   Modified $default_graph for new Linear
#                                       New $date_format parameter
#  J. Loiacono  01/26/2014      4.3     Introduced Detect Scanning
#  J. Loiacono  07/04/2014      4.4     Multiple dashboards and Analysis
#  J. Loiacono  11/02/2014      4.5     New $silk_compiled_localtime
#                                       New $ipset_directory
#                                       New $use_bottom_pulldowns
#                                       New $ipfix_default_device
#                                       New $sensor_config_file [rename]
#                                       New $silk_config_file
#                                       FlowTracker to FlowMonitor name changes 
#  J. Loiacono  01/26/2015      4.6     Removed $time_zone (now from system)
#
#$Author$
#$Date$
#$Header$
#
###########################################################################
#
#               BEGIN EXECUTABLE STATEMENTS
#
 
# Path variable

$ENV{PATH}              .= ':/usr/local/bin:/usr/sbin';

# Server

$FlowViewer_server       = "192.168.100.1"; # (IP address or hostname)

# Service

$FlowViewer_service      = "https";           # (http, or https)

# Directories and Files:

$reports_directory       = "/var/www/html/FlowViewer";
$reports_short           = "/FlowViewer";
$graphs_directory        = "/var/www/html/FlowGrapher";
$graphs_short            = "/FlowGrapher";
$monitor_directory       = "/var/www/html/FlowMonitor";
$monitor_short           = "/FlowMonitor";
$cgi_bin_directory       = "/var/www/cgi-bin/FlowViewer_4.6";
$cgi_bin_short           = "/cgi-bin/FlowViewer_4.6";
$work_directory          = "/var/www/cgi-bin/FlowViewer_4.6/Flow_Working";
$save_directory          = "/var/www/html/FlowViewer_Saves";
$save_short              = "/FlowViewer_Saves";
$names_directory         = "/var/www/cgi-bin/FlowViewer_4.6";
$ipset_directory         = "/var/www/cgi-bin/FlowViewer_4.6";     # Where FlowViewer can find IPset files
$filter_directory        = "/var/www/cgi-bin/FlowMonitor_Files/FlowMonitor_Filters";  
$rrdtool_directory       = "/var/www/cgi-bin/FlowMonitor_Files/FlowMonitor_RRDtool";
$dashboard_directory     = "/var/www/html/FlowViewer_Dashboard";
$dashboard_short         = "/FlowViewer_Dashboard";
#@other_dashboards       = ();          # Set to empty if you have just the one nominal Dashboard
@other_dashboards        = ("/var/www/html/SOC","/var/www/html/NetOps");
#@dashboard_titles       = ();          # Set to empty if you have just the one nominal Dashboard
@dashboard_titles        = ("Performance","SOC","NetOps"); # titles must be in the same order as the directories

$flow_data_directory     = "/data/flows";
$exporter_directory      = "/data/flows/all_routers";
$flow_bin_directory      = "/usr/local/flow-tools/bin";
$rrdtool_bin_directory   = "/usr/bin";

# SiLK parameters

$silk_data_directory     = "/data/flows";
$silk_bin_directory      = "/usr/local/bin";
$site_config_file        = "/data/flows/silk.conf";          # If left blank, will look for silk.conf in specified Data Rootdir (see User's Guide)
$sensor_config_file      = "/data/flows/sensor.conf";
$silk_compiled_localtime = "";          # Set to "Y" if you compiled SiLK with --enable-localtime switch

$silk_capture_buffer_pre = (125 * 60);  # Start of SiLK file concatenation
$silk_capture_buffer_post= (5 * 60);    # End of SiLK file concatenation

$silk_init_loadscheme    = 1;           # For Flows Initiated/Second - see SiLK rwcount documentation
$silk_active_loadscheme  = 5;           # For Flows Active/Second - see SiLK rwcount documentation
$silk_class_default      = "";          # General SiLK file structure info. silk.conf, sensor.conf
$silk_flowtype_default   = "";          # General SiLK file structure info. silk.conf, sensor.conf
$silk_type_default       = "all";       # General SiLK file structure info. silk.conf, sensor.conf
$silk_sensors_default    = "";          # General SiLK file structure info. silk.conf, sensor.conf
$silk_switches_default   = "";          # General SiLK file structure info. silk.conf, sensor.conf

# General parameters

$version                 = "4.6";
$no_devices_or_exporters = "N";
@devices                 = ("router_1","router_2","router_3","router_4","router_5","router_6"); # for flow-tools
@ipfix_devices           = ("router_ipfix_1","router_ipfix_2","Site");   # for SiLK, if none: @ipfix_devices = ();
@ipfix_storage           = ("router_ipfix_1:15G","router_ipfix_2:10G");  # If using FlowViewer_CleanSilk, set to storage requirements for each device
$ipfix_default_device    = "";          # All initial, blank forms will have this selected instead of "Select Device"
#@exporters              = ("192.168.200.1:New York Router","192.168.200.2:Prague Router");
@exporters               = ();

$flow_capture_interval   = (35 * 60);
$flow_file_length        = (15 * 60);
$start_offset            = (90 * 60);   # e.g., 90 minutes ago
$end_offset              = (30 * 60);   # e.g., 30 minutes ago
$use_even_hours          = "Y";
$N                       = 3;
$use_NDBM                = "N";
$pie_chart_default       = 0;           # 0 = None;  1 = With Others;  2 = Without Others
$number_slices           = 9;
$pie_colors              = ['pie2 color1','pie2 color2','pie2 color3','pie2 color4','pie2 color5','pie2 color6','pie2 color7','pie2 color8','pie2 color9','pie2 color10'];
$maximum_days            = "91";
$remove_workfiles_time   = 2*86400;
$remove_graphfiles_time  = 7*86400;
$remove_reportfiles_time = 7*86400;
$time_zone_dst_offset    = (60 * 60);   # Number of seconds of the Daylight Savings adjustment in your timezone
$date_format             = "MDY";       # MDY=MM/DD/YYYY  DMY=DD/MM/YYYY  DMY2=DD.MM.YYYY  YMD=YYYY-MM-DD
$labels_in_titles        = "1";         # Set to "1" for labels in Monitor graph titles; "0" off
$sip_prefix_length       = "16";
$dip_prefix_length       = "16";

# UI Parameters

$left_title              = "SWAN Network";
$left_title_link         = "$cgi_bin_short/FV.cgi";
$right_title             = "Monitoring SWAN Network Data Flows";
$right_title_link        = "$cgi_bin_short/FV.cgi";
$use_bottom_pulldowns    = "Y";

# Debug Parameters 

$debug_viewer            = "Y";
$debug_grapher           = "Y";
$debug_monitor           = "Y";
$debug_group             = "Y";
$debug_files             = "N";

# Graphing parameters
 
$transparent        = "0";   
$x_ticks            = "T";
$long_ticks         = "T";
$skip_undef         = "T";
$graph_height       = 310;
$graph_width        = 600;
$t_margin           = 10;
$b_margin           = 60;
$l_margin           = 10;
$r_margin           = 20;
$bgclr              = "white";
$borderclrs         = "black";
$boxclr             = "white";
$fgclr              = "gray90";
$labelclr           = "black";
$axislabelclr       = "black";
$legendclr          = "black";
$valuesclr          = "black";
$textclr            = "black";
$x_axis_font        = "('arial', 16)";
$title_font         = "('arial', 18)";
$horz_max           = ($graph_width / 2) - 44;
$horz_pct           = ($graph_width / 2) - 44;
$horz_avg           = ($graph_width / 2) - 44;
$horz_min           = ($graph_width / 2) - 44;
$vert_max           = ($graph_height - 70) + 2;
$vert_pct           = ($graph_height - 70) + 16;
$vert_avg           = ($graph_height - 70) + 30;
$vert_min           = ($graph_height - 70) + 44;
$horz_mth           = 15;
$analyze_count      = 8;         # Any number from 3 to 10 inclusive. Must have at least [ $analyze_count+1 ] analyze_colors
$analyze_peak_width = 1000;      # Number of observations to examine for peaks (per period)
#$analyze_colors    = ['gray95','pale green','pale brown','pale red','pale blue','pale yellow'];
#$analyze_colors    = ['gray95','pastel orange','pastel rose','pastel blue','pastel green','pastel yellow'];
#$analyze_colors    = ['gray95','pie2 color1','pie2 color2','pie2 color3','pie2 color4','pie2 color5','pie2 color6','pie2 color7','pie2 color8','pie2 color9','pie2 color10'];
#$analyze_colors    = ['gray95','auto mixed1','auto mixed2','auto mixed3','auto mixed4','auto mixed5','auto mixed6','auto mixed7','auto mixed8','auto mixed9','auto mixed10'];
#$analyze_colors    = ['gray95','analysis1','analysis2','analysis4','analysis5','analysis9','auto mixed1'];
$analyze_colors     = ['gray95','analysis1','analysis2','analysis3','analysis4','analysis5','analysis6','analysis7','analysis8','analysis9','analysis10'];
$analyze_extension  = 20;        # Number of pixels to extend FlowGrapher_Analyze graph height

# Monitor parameters
 
$actives_webpage    = "index.html";
$log_directory      = "/var/www/cgi-bin/FlowViewer_4.6/logs";
$log_collector_short= "Y";
$log_collector_med  = "N";
$log_collector_long = "N";
$log_grapher_short  = "Y";
$log_grapher_long   = "N";
$collection_offset  = 1800;
$collection_period  = 300;
$graphing_period    = 300;
$recreate_cat_length= 6*(60*60); # Time length of concatenated file

$rrd_dir_perms      = 0777;      # Scale these back once everything is working 
$filter_dir_perms   = 0777;
$work_dir_perms     = 0777;
$html_dir_perms     = 0777;

$html_file_perms    = 0777;
$graph_file_perms   = 0777;
$rrd_file_perms     = 0777;
$filter_file_perms  = 0777;
$monitor_file_perms = 0777;
$saved_filters_perms= 0777;
$actives_file_perms = 0777;

$rrd_area           = "FFE0C0";   
$rrd_line           = "000000";   
$rrd_peak           = "000000";
$rrd_width          = 600;
$rrd_height         = 150;
$rrd_font           = "000000AA";
$rrd_back           = "FFFFFF";
$rrd_canvas         = "FFFFFF";
$rrd_grid           = "CCCCCC88";
$rrd_mgrid          = "FF000033";
$rrd_frame          = "FFFFFF";
$rrd_shadea         = "FFFFFF";
$rrd_shadeb         = "FFFFFF";
$rrd_thick          = 0.3;
$rrd_lower_limit    = 0;
$rrd_slope_mode     = "--slope-mode";      # $rrd_slope_mode = ""; will square graphs up
$rrd_vrule_color    = "FF0000";
$rrd_hrule_color    = "FF0000";
$thumbnail_width    = 250;
$thumbnail_height   = 80;
$hr_width           = $rrd_width + 130;

# Standard Deviation Alert parameters

$sigma_type_1       = "6:2.67";   # Num of obs in mean : number of sigmas for threshold : Must restart FlowMonitor_Collector
$sigma_type_2       = "12:4";     # Num of obs in mean : number of sigmas for threshold : Must restart FlowMonitor_Collector
$sigma_type_3       = "12:3";     # Num of obs in mean : number of sigmas for threshold : Must restart FlowMonitor_Collector

# Scanning Parameters

$dscan_parameters   = "-w -W";    # flow-tools only, ignores inbound and outbound port 80
$scan_model         = "2";        # SiLK only: 0=TRW&BLR; 1=TRW only;  2=BLR only
$trw_internal_set   = "$cgi_bin_directory/edc_ipset.set";       # SiLK only: file_name, required when using TRW model

# Webpage Parameters

$filename_color     = "#CF7C29";
$dns_column_width   = 60;
$detail_lines       = 200;
$asn_width          = 60;
$default_report     = 10;         # See FlowViewer Users Guide for details
$default_graph      = "bps";      # See FlowViewer Users Guide for details
$default_lines      = 100;
$default_identifier = "DNS";      # Use "IP" for IP addresses; "DNS" to resolve addresses to names
$default_flows      = 1;          # See FlowViewer Users Guide for details

# Commands (full directory names)
 
$dig                = "/usr/bin/dig +time=1 +tries=1 -x ";
$dig_forward        = "/usr/bin/dig +time=1 +tries=1 ";
